hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/MultipartCryptoKeyInputStream.java

```diff
@@ -0,0 +1,373 @@
+            numBytesRead);
+
+        LOG.debug("OzoneCryptoInputStream for key: {} part: {} read {} bytes " +
+                "instead of {} bytes to account for Crypto buffer boundary. " +
+                "Client buffer will be copied with read data from position {}" +
+                "upto position {}, discarding the extra bytes read to " +
+                "maintain Crypto buffer boundary limits", key, partIndex,
+            actualNumBytesRead, numBytesRead, readPositionAdjustedBy,
+            actualNumBytesRead - readPositionAdjustedBy);
+
+        // Reset readPositionAdjustedBy and readLengthAdjustedBy
+        readPositionAdjustedBy = 0;
+        readLengthAdjustedBy = 0;
+      } else {
+        numBytesRead = current.read(b, off, numBytesToRead);
+        if (numBytesRead != numBytesToRead) {
+          throw new IOException(String.format("Inconsistent read for key=%s " +
+                  "part=%d length=%d numBytesToRead=%d numBytesRead=%d",
+              key, partIndex, current.getLength(), numBytesToRead,
+              numBytesRead));
+        }
+      }
+
+      totalReadLen += numBytesRead;
+      off += numBytesRead;
+      len -= numBytesRead;
+
+      if (current.getRemaining() <= 0 &&
+          ((partIndex + 1) < partStreams.size())) {
+        partIndex += 1;
+      }
+
+    }
+    return totalReadLen;
+  }
+
+  /**
+   * Get number of bytes to read from the current stream based on the length
+   * to be read, number of bytes remaining in the stream and the Crypto buffer
+   * size.
+   * Reads should be performed at the CryptoInputStream Buffer boundaries only.
+   * Otherwise, the decryption will be incorrect.
+   */
+  private int getNumBytesToRead(int lenToRead, int remaining,
+      int cryptoBufferSize) throws IOException {
+
+    Preconditions.checkArgument(readPositionAdjustedBy == 0);
+    Preconditions.checkArgument(readLengthAdjustedBy == 0);
+
+    // Check and adjust position if required
+    adjustReadPosition(cryptoBufferSize);
+    remaining += readPositionAdjustedBy;
+    lenToRead += readPositionAdjustedBy;
+
+    return adjustNumBytesToRead(lenToRead, remaining, cryptoBufferSize);
+  }
+
+  /**
+   * Reads should be performed at the CryptoInputStream Buffer boundary size.
+   * Otherwise, the decryption will be incorrect. Hence, if the position is
+   * not at the boundary limit, we have to adjust the position and might need
+   * to read more data than requested. The extra data will be filtered out
+   * before returning to the client.
+   */
+  private void adjustReadPosition(long cryptoBufferSize) throws IOException {
+    // Position of the buffer in current stream
+    long currentPosOfStream = partStreams.get(partIndex).getPos();
+    if (currentPosOfStream % cryptoBufferSize != 0) {
+      // Adjustment required.
+      // Update readPositionAdjustedBy and seek to the adjusted position
+      readPositionAdjustedBy = (int) (currentPosOfStream % cryptoBufferSize);
```

Comment:
NIT: readPositionAdjustedBy = (int) (currentPosOfStream % cryptoBufferSize);
can be moved before if to save a % operation.